[RFC FS-1075] Improve interop to C# nullable-typed optionals

dsyme opened the pull request:
Implementation of [F# RFC FS-1075 - Interop with C# nullable-typed optional parameters](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1075-nullable-interop.md)

* [x] use feature branch
* [x] language version protection

Continues https://github.com/dotnet/fsharp/pull/7296

dsyme commented:
I added the language version protection and tests for it, so I think this is actually all ready for review

TIHan commented:
Great! Will look at it.

abelbraaksma commented:
@dsyme, the discussion link on the spec returns a 404. Do we have a dedicated discussion thread for this RFC? Could you update?

cartermp commented on src/fsharp/IlxGen.fs:
```diff
@@ -5089,7 +5089,7 @@ and GenDecisionTreeSwitch cenv cgbuf inplabOpt stackAtTargets eenv e cases defau
 
 and GenDecisionTreeCases cenv cgbuf stackAtTargets eenv defaultTargetOpt targets repeatSP targetInfos sequel caseLabels cases (contf: Zmap<_,_> -> FakeUnit) =
-    assert(cgbuf.GetCurrentStack() = stackAtTargets) // cgbuf stack should be unchanged over tests. [bug://1750].
+    //assert(cgbuf.GetCurrentStack() = stackAtTargets) // cgbuf stack should be unchanged over tests. [bug://1750].
```
Why is this commented out?

dsyme replied:
The assert is triggering in debug builds and is just not accurate

smoothdeveloper replied:
Confirming I've hit this assert a bunch of times recently.

cartermp commented on src/fsharp/LanguageFeatures.fs:
```diff
@@ -28,6 +28,7 @@ type LanguageFeature =
     | NameOf = 6
     | ImplicitYield = 7
     | OpenStaticClasses = 8
+    | NullableOptionalInterop = 1075 // number is RFC
```
I think we can just do away with the numbered things; I'll submit a PR

TIHan commented on src/fsharp/MethodCalls.fs:
```diff
@@ -42,13 +43,18 @@ open FSharp.Compiler.ExtensionTyping
     | CallerArg of TType * range * bool * 'T  
-    member x.Type = (let (CallerArg(ty, _, _, _)) = x in ty)
+
+    member x.CallerArgumentType = (let (CallerArg(ty, _, _, _)) = x in ty)
```
nit: The type is called `CallerArg` and this property on it is called `CallerArgumentType` which feels a little redundant, I feel `Type` was more appropriate.

dsyme replied:
We generally qualify the name "type", e.g. "PropertyType" on "PropertyInfo", "FieldType" on "FieldInfo" etc.    I think it's good practice given how many meanings the word carries

smoothdeveloper replied:
FYI in this PR https://github.com/dotnet/fsharp/pull/6596/files#diff-456bef6dcdc9277074038269046cab70 I'm changing this type to a record, it didn't impact too much code.

I also introduce `CallerArgs` to replace `curriedCallerArgs: (CallerArg<'T> list * CallerNamedArg<'T> list) list,` in `type CalledMeth<'T>` constructor.

I'd like to understand why those values end up being list of tuple of lists instead of just a tuple of lists? It seems to run pretty deep up to `SynValInfo` but there is no indication of what those nested lists are.

@dsyme: could you cast some light?

TIHan replied:
@dsyme ok, will keep that in mind. :)

dsyme commented:
@cartermp This feature is ready and I think can be integrated into preview releases

KevinRansom commented on src/fsharp/MethodCalls.fs:
```diff
@@ -1183,6 +913,408 @@ let BuildMethodCall tcVal g amap isMutable m isProp minfo valUseFlags minst objA
+   if isByrefTy g calledArgTy && isRefCellTy g callerArgTy then 
+       None, Expr.Op (TOp.RefAddrGet false, [destRefCellTy g callerArgTy], [callerArgExpr], m) 
+
+#if IMPLICIT_ADDRESS_OF
```
IMPLICIT_ADDRESS_OF is never actually set, is there any reason to keep this?

TIHan replied:
It's never set. I added this to document where the implicit address-of feature for methods with inref argument types can get activated. We never fully fleshed out the rest of that feature. Perhaps we could remove this and make a feature branch on it.

KevinRansom commented on src/fsharp/MethodCalls.fs:
```diff
@@ -1183,6 +913,408 @@ let BuildMethodCall tcVal g amap isMutable m isProp minfo valUseFlags minst objA
+                        let callerArgExprCoerced = mkCoerceIfNeeded g calledNonOptTy callerArgTy callerArgExpr
+                        mkSome g calledNonOptTy callerArgExprCoerced m
+                    else 
+                        callerArgExpr // should be unreachable 
```
Is the "should be unreachable" in the above line an invariant or an aspiration?
probably should be a failwith or remove the comment.

dsyme replied:
thanks, I added an assert to the branch

TIHan commented on src/fsharp/MethodCalls.fs:
```diff
@@ -96,14 +106,104 @@ type AssignedItemSetter<'T> = AssignedItemSetter of Ident * AssignedItemSetterTa
+
+        | NotOptional -> 
+            // This condition is caught later
+            //warning(InternalError("Unexpected NotOptional", m))
```
Is it fair we remove this comment?

dsyme commented:
@TIHan 
> After testing the feature out, there is existing F# code that can break once this feature is enabled, ex:

I've fixed this problem, thanks for finding it!

This requires the addition of two new rules to the RFC

* when comparing overloads, and overload argument of type `X` is preferred to one of type `Nullable<X>` if they otherwise both match

* when comparing overloads, we were only comparing unnamed arguments.  Now, if two overloads were considered equal priority by previous rules (i.e. comparison returned zero), then their entire argument lists are compared including both unnamed and named arguments, using the same rules as already used for unnamed arguments.  We use this as a last-resort comparison.

I've updated the PR and will update the RFC

dsyme commented:
The update to the RFC is here (I did it via direct commit)

https://github.com/fsharp/fslang-design/commit/43785ddbeb013ff3635941fb417b6e0ed46f7582#diff-67995b3eec7592c0a05c96bff39a7a83

That resolves all issues with the PR as it stands, please re-review?  @TIHan Thanks!

smoothdeveloper commented on src/fsharp/MethodCalls.fs:
```diff
@@ -445,7 +514,7 @@ type CalledMeth<'T>
     member x.ParamArrayCallerArgs = x.ArgSets |> List.tryPick (fun argSet -> if Option.isSome argSet.ParamArrayCalledArgOpt then Some argSet.ParamArrayCallerArgs else None )
 
-    member x.ParamArrayElementType = 
+    member x.GetParamArrayElementType() = 
```
Curious if you changed that to a method due to the assert showing in variable inspector during debug or if it is more stylistic about the thing not being a property?

dsyme replied:
> Curious if you changed that to a method due to the assert showing in variable inspector during debug or if it is more stylistic about the thing not being a property?

For the former reason

TIHan commented:
Ok, will look at this again. Thank you!
